compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -217,90 +221,91 @@ class ClassfileParser(
     }
   }
 
-  def parseMember(method: Boolean)(using Context): Unit = {
+  def parseMember(method: Boolean)(using ctx: Context, in: DataReader): Unit = {
     val start = indexCoord(in.bp)
     val jflags = in.nextChar
     val sflags =
       if (method) Flags.Method | methodTranslation.flags(jflags)
       else fieldTranslation.flags(jflags)
-    val name = pool.getName(in.nextChar)
-    if (!sflags.isOneOf(Flags.PrivateOrArtifact) || name == nme.CONSTRUCTOR) {
+    val preName = pool.getName(in.nextChar)
+    if (!sflags.isOneOf(Flags.PrivateOrArtifact) || preName.name == nme.CONSTRUCTOR) {
+      val sig = pool.getExternalName(in.nextChar).value
+      val completer = MemberCompleter(preName.name, jflags, sig)
       val member = newSymbol(
-        getOwner(jflags), name, sflags, memberCompleter,
+        getOwner(jflags), preName.name, sflags, completer,
         getPrivateWithin(jflags), coord = start)
+
+      completer.attrCompleter = parseAttributes(member)
+
       getScope(jflags).enter(member)
+
+    }
+    else {
+      in.nextChar // info
+      skipAttributes()
     }
-    // skip rest of member for now
-    in.nextChar // info
-    skipAttributes()
   }
 
-  val memberCompleter: LazyType = new LazyType {
+  class MemberCompleter(name: SimpleName, jflags: Int, sig: String) extends LazyType {
```

Comment:
I suggest moving this class outside of the ClassfileParser class to make sure we don't accidentally capture any reference to the class or one of its field.

Reply:
That is a good refactoring, but unfortunately it indirectly depends on `innerClasses` and some other member methods. We have to do some big operations for classfile parser class. I'm happy to do such a refactoring to make things more clear in the classfile parser in another PR, and make this PR simpler and single-purpose.